holoviews/element/raster.py

```diff
@@ -408,45 +211,157 @@ class Image(SheetCoordinateSystem, Raster):
-        l, b, r, t = bounds.lbrt()
-        extents = extents if extents else (None, None, None, None)
-        Element2D.__init__(self, data, extents=extents, bounds=bounds,
-                           **params)
 
-        (dim1, dim2) = self.data.shape[1], self.data.shape[0]
+        l, b, r, t = bounds.lbrt()
         xdensity = xdensity if xdensity else dim1/float(r-l)
         ydensity = ydensity if ydensity else dim2/float(t-b)
         SheetCoordinateSystem.__init__(self, bounds, xdensity, ydensity)
 
-        if len(self.data.shape) == 3:
-            if self.data.shape[2] != len(self.vdims):
+        if len(self.shape) == 3:
+            if self.shape[2] != len(self.vdims):
                 raise ValueError("Input array has shape %r but %d value dimensions defined"
-                                 % (self.data.shape, len(self.vdims)))
+                                 % (self.shape, len(self.vdims)))
+
+
+    def __setstate__(self, state):
+        """
+        Ensures old-style unpickled Image types without an interface
+        use the ImageInterface.
+        """
+        self.__dict__ = state
+        if isinstance(self.data, np.ndarray):
+            self.interface = ImageInterface
+        super(Dataset, self).__setstate__(state)
+
+
+    def aggregate(self, dimensions=None, function=None, spreadfn=None, **kwargs):
+        agg = super(Image, self).aggregate(dimensions, function, spreadfn, **kwargs)
+        return Curve(agg) if isinstance(agg, Dataset) and len(self.vdims) == 1 else agg
+
+
+    def select(self, selection_specs=None, **selection):
+        """
+        Allows selecting data by the slices, sets and scalar values
+        along a particular dimension. The indices should be supplied as
+        keywords mapping between the selected dimension and
+        value. Additionally selection_specs (taking the form of a list
+        of type.group.label strings, types or functions) may be
+        supplied, which will ensure the selection is only applied if the
+        specs match the selected object.
+        """
+        if selection_specs and not any(self.matches(sp) for sp in selection_specs):
+            return self
 
+        selection = {self.get_dimension(k).name: slice(*sel) if isinstance(sel, tuple) else sel
+                     for k, sel in selection.items() if k in self.kdims}
+        coords = tuple(selection[kd.name] if kd.name in selection else slice(None)
+                       for kd in self.kdims)
 
-    def _convert_element(self, data):
-        if isinstance(data, (Raster, HeatMap)):
-            return data.data
+        indexed = False
+        if not any([isinstance(el, slice) for el in coords]):
+            indexed = True
+            selection = {kd.name: c for kd, c in zip(self.kdims, self.closest(coords))}
+
+        # Compute new bounds
+        shape = self.interface.shape(self)
+        ys, xs = shape[:2]
+        xidx, yidx = coords
+        xdim, ydim = self.kdims
```

Comment:
Looks like xdim and ydim could be moved (not used immediately here).

Could ``compute_new_bounds`` be a utility? 

Something like ``compute_new_bounds(shape, coords, lbrt, xdensity, ydensity)``